### DISCUSSION TITLE
BRouter no recalculation possible on the highway in case of "avoid highway"

### DISCUSSION THREAD

**Discussion by WolfgangZ42** [devemux86/cruiser#188] [Q&A] [answered]
Based on this small route and "avoid highway" I compared Kurviger and Cruiser:
<img width="500" src="https://github.com/devemux86/cruiser/assets/121806238/68f4425a-1b69-4187-be1c-178cb5bdd28c"/>

If I nevertheless drive on the highway (A94), the recalculation of K2 recommends me to leave the highway at the next exit. 

With Cruiser a recalculation is simply not possible. The whole thing is caused by BRouter, because if I set the routing service in K2 to BRouter alone, the reaction is the same as from Cruiser. 

Obviously BRouter considers the not allowed highway as non-existent:
<img width="500" src="https://github.com/devemux86/cruiser/assets/121806238/9171509a-470a-41e5-8ce7-e9e380e858f1"/>

Maybe this can be optimized?

**Comment by devemux86** [comment]
@WolfgangZ42 thanks for the report!

Can you share a [BRouter link](https://brouter.de/brouter-web/) with this route?

**Comment by MotoUKRider** [comment]
I can replicate this with my local roads - the same occurs if there's a Shaping-Point on the motorway: I get the `"not mapped in existing datafile"` message.  There is an opportunity to leave the motorway at the next exit, but as @WolfgangZ42 says, it needs "optimization" in that it should be able to take you *off* the highway, but not route you *onto* it :)

**Comment by devemux86** [comment]
Here is a [BRouter link](https://brouter.de/brouter-web/#map=15/48.1681/11.8806/standard&lonlats=11.866738,48.162565;11.943686,48.226741&profile=car-fast) for the above example.

BRouter seems to have absolute avoidances, Kurviger has lenient avoidances.
So you mean: if there is no other way, take the motorway until we can exit it?

With BRouter profiles we can change everything! :+1: 

Please try the car profile on BRouter website, enable the "avoid_toll" and "avoid_motorways" and it will not route.

But if we change line 188, to e.g.:
```
  min if is_avoided_motorway then 1 else 999
```
Then it works: it calculates the route and exits the motorway quickly on the 1st exit.

**Reply by WolfgangZ42** [reply]
> With BRouter profiles we can change everything! 👍

Unbelievable - my deep respect for the possibilities BRouter offers!

**Reply by DolomitesTom** [reply]
perfect, this should be the default for Cruiser

**Reply by Ciutnik** [reply]
> But if we change line 188, to e.g.:
`  min if is_avoided_motorway then 1 else 999`
Then it works: it calculates the route and exits the motorway quickly on the 1st exit.

It works great! An absolutely mandatory setting. I wonder if it is possible to similarly limit – but not completely exclude – expressways and primary roads?

**Reply by WolfgangZ42** [reply]
> But if we change line 188, to   _min if is_avoided_motorway then 1 else 999_    it works!

Today - still with 3.0.20:
<img width="400" src="https://github.com/devemux86/cruiser/assets/121806238/869cd865-0b43-4428-be4d-d5e6509e6137"/>
Cruiser on the way to the nearest exit off the "forbidden" highway ... 😉😂

Question: is  _min if is_avoided_motorway then 1 else 999_  implemented in car-cruiser (the internal .brf file)?

**Reply by devemux86** [reply]
> Question: is min if is_avoided_motorway then 1 else 999 implemented in car-cruiser (the internal .brf file)?

The modification for `is_avoided_motorway` will be included in the next Cruiser 3.0.21.
We can also do the same for `is_avoided_toll_road`.

Cruiser 3.0.20 was already released when I examined this report. :slightly_smiling_face:

**Reply by devemux86** [reply]
Here is the next [car-cruiser.brf.zip](https://github.com/devemux86/cruiser/files/11604562/car-cruiser.brf.zip) planned for next Cruiser,
if anyone wants to test it.

**Reply by Ciutnik** [reply]
Maybe it would be useful, and probably easy to implement, to introduce a defined approximate avoidance level for these parameters?

**Reply by devemux86** [reply]
> I wonder if it is possible to similarly limit – but not completely exclude – expressways and primary roads?

> introduce a defined approximate avoidance level for these parameters?

@Ciutnik thanks for the ideas!
Indeed, these would be very useful additions, I will need to examine them.

**Comment by DolomitesTom** [comment]
My Cruiser specific brf files have been deleted, where are they now after the last update?

**Reply by devemux86** [reply]
@DolomitesTom sorry about that, it was a side effect of the latest Cruiser changes for BRouter... :slightly_smiling_face: 

- https://github.com/devemux86/cruiser/discussions/175#discussioncomment-6039934

**Reply by DolomitesTom** [reply]
Ah ok, is there now only 1 profile necessary? The internal or 1 external?

**Reply by DolomitesTom** [reply]
What means use tracks?

**Reply by devemux86** [reply]
We can open an external BRouter profile (located anywhere) or use the internal profiles.
This is the mapping (currently only `car-cruiser` is modified):

|Cruiser|BRouter|
|-|-|
|Car / Motorcycle|car-cruiser|
|Bike|fastbike|
|Mountain bike|trekking|
|Walking / Hiking|hiking-mountain|

**Reply by devemux86** [reply]
> What means use tracks?

Please see the answer here:
- https://github.com/devemux86/cruiser/discussions/180#discussioncomment-6040390

**Comment by devemux86** [comment, marked as answer]
New BRouter profile with non-absolute avoid for motorways and toll roads is available in [Cruiser 3.0.22](https://play.google.com/store/apps/details?id=gr.talent.cruiser).
